### PR TITLE
refactor(e2e): use `libvmi` instead of `NewRandomVMIWithDataVolume`

### PR DESCRIPTION
**Pull request by fossedihelm**
### What this PR does
Before this PR:
`tests/utils.go` file contains several utility functions that can be considered duplicated.
This is because the same result can be achieved with the already present libvmi/etc options.

After this PR:
This PR aims to drop the `NewRandomVMIWithDataVolume` wrapper function,
replacing its occurrences with:
```
vmi := libvmi.New(
	libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
	libvmi.WithNetwork(v1.DefaultPodNetwork()),
	libvmi.WithDataVolume("disk0", dataVolume.Name),
	libvmi.WithResourceMemory("1Gi"),
)
```

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer

During the conversion, some other improvements have been made, in the vision of using `libvmi.Option`.
In particular:
- `tests.AddUserData(vmi, ...)` --> `libvmi.WithCloudInitNoCloudEncodedUserData(...),`
- `libstorage.AddDataVolumeTemplate(vm, dataVolume)` --> `libvmi.WithDataVolumeTemplate(dataVolume))`
- `addDataVolumeDisk(vm, "blankdisk", blankDv.Name)` --> `libvmi.WithDataVolume("blankdisk", blankDv.Name)`
- `vm.Spec.Running = pointer.Bool(true)` --> `libvmi.WithRunning()`
- `vmi.ObjectMeta.Annotations = renderSidecar()` --> `libvmi.WithAnnotation("hooks.kubevirt.io/hookSidecars", hookSidecarsValue),...`
- `vmi1.Spec.Affinity = affinityRule --> `libvmi.WithNodeAffinityForLabel(labelKey, ""),`
- `vmi1.Labels = labels` --> `libvmi.WithLabel(labelKey, ""),`
					

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] ~Upgrade: Impact of this change on upgrade flows was considered and addressed if required~
- [ ] ~Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test~
- [ ] ~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

### Release note

```release-note
NONE
```

### CODE REVIEW

File: tests/storage/storage.go [hunk 1]
```diff
@@ -1126,39 +1126,23 @@ var _ = SIGDescribe("Storage", func() {
...
 
-				vmi1.Spec.Affinity = affinityRule
-				vmi2.Spec.Affinity = affinityRule
+				vmi1 = libvmi.New(
+					libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+					libvmi.WithNetwork(v1.DefaultPodNetwork()),
+					libvmi.WithDataVolume("disk0", dv.Name),
+					libvmi.WithResourceMemory("1Gi"),
+					libvmi.WithLabel(labelKey, ""),
+					libvmi.WithNodeAffinityForLabel(labelKey, ""),
```
Comment (xpivarc) [inline tests/storage/storage.go:1136]: This is not the same, I have not checked if this is ok but I would ask if we have equivalent.

Comment (fossedihelm) [reply tests/storage/storage.go:1136]: You are right! Honestly, I missed the pod/node affinity! As is written right now, it is not equivalent, and will cause flakiness IMHO.
Thanks for catching this

File: tests/libstorage/datavolume.go [hunk 2]
```diff
@@ -43,27 +43,6 @@ import (
 	"kubevirt.io/kubevirt/tests/util"
 )
 
-func AddDataVolumeDisk(vmi *v13.VirtualMachineInstance, diskName, dataVolumeName string) *v13.VirtualMachineInstance {
```
Comment (0xFelix) [inline tests/libstorage/datavolume.go:46]: Not related: `v13` is really a weird name for the kubevirt import.

File: tests/migration/migration.go [hunk 3]
```diff
@@ -2196,8 +2211,13 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
...
 				Expect(err).ToNot(HaveOccurred())
 				libstorage.EventuallyDV(dv, 600, HaveSucceeded())
-				vmi := tests.NewRandomVMIWithDataVolume(dv.Name)
-				tests.AddUserData(vmi, "disk1", "#!/bin/bash\n echo hello\n")
+				vmi := libvmi.New(
+					libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+					libvmi.WithNetwork(v1.DefaultPodNetwork()),
+					libvmi.WithDataVolume("disk0", dv.Name),
+					libvmi.WithResourceMemory("1Gi"),
+					libvmi.WithCloudInitNoCloudEncodedUserData("#!/bin/bash\n echo hello\n"),
```
Comment (0xFelix) [inline tests/migration/migration.go:2219]: Make this a constant in libvmi factory.go?

https://github.com/kubevirt/kubevirt/blob/9d4a040652074af39d58eeea6a9cf3042b64d285/tests/libvmi/factory.go#L51

File: tests/storage/datavolume.go [hunk 4]
```diff
@@ -1196,10 +1229,14 @@ var _ = SIGDescribe("DataVolume Integration", func() {
...
 			preallocated := dataVolume.Spec.Preallocation != nil && *dataVolume.Spec.Preallocation
 
-			vmi := tests.NewRandomVMIWithDataVolume(dataVolume.Name)
-			vmi.Spec.Domain.Resources.Requests[k8sv1.ResourceMemory] = resource.MustParse("512M")
+			vmi := libvmi.New(
+				libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+				libvmi.WithNetwork(v1.DefaultPodNetwork()),
+				libvmi.WithDataVolume("disk0", dataVolume.Name),
+				libvmi.WithResourceMemory("512M"),
+				libvmi.WithCloudInitNoCloudEncodedUserData("#!/bin/bash\n echo hello\n"),
```
Comment (0xFelix) [inline tests/storage/datavolume.go:1237]: Same as above

File: tests/storage/datavolume.go [hunk 5]
```diff
@@ -1461,10 +1498,14 @@ func newRandomVMWithCloneDataVolume(sourceNamespace, sourceName, targetNamespace
...
+	vmi := libvmi.New(
+		libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+		libvmi.WithNetwork(v1.DefaultPodNetwork()),
+		libvmi.WithDataVolume("disk0", dataVolume.Name),
+		libvmi.WithResourceMemory("1Gi"),
+		libvmi.WithNamespace(targetNamespace),
+	)
 
-	libstorage.AddDataVolumeTemplate(vm, dataVolume)
+	vm := libvmi.NewVirtualMachine(vmi, libvmi.WithDataVolumeTemplate(dataVolume))
```
Comment (0xFelix) [inline tests/storage/datavolume.go:1509]: You inlined the call to `libvmi.New` above, do it here too?

File: tests/storage/export.go [hunk 6]
```diff
@@ -1867,11 +1879,16 @@ var _ = SIGDescribe("Export", func() {
...
+		vmi := libvmi.New(
+			libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+			libvmi.WithNetwork(virtv1.DefaultPodNetwork()),
+			libvmi.WithDataVolume("disk0", dataVolume.Name),
+			libvmi.WithResourceMemory("1Gi"),
+			libvmi.WithCloudInitNoCloudEncodedUserData(bashHelloScript),
+			libvmi.WithDataVolume("blankdisk", blankDv.Name),
+			libvmi.WithNamespace(testsuite.GetTestNamespace(nil)),
+		)
+		vm := libvmi.NewVirtualMachine(vmi, libvmi.WithRunning())
```
Comment (0xFelix) [inline tests/storage/export.go:1891]: Same as above.

File: tests/storage/hotplug.go [hunk 7]
```diff
@@ -1322,9 +1313,17 @@ var _ = SIGDescribe("Hotplug", func() {
...
-				vmi.ObjectMeta.Annotations = renderSidecar()
-				tests.AddUserData(vmi, "cloud-init", "#!/bin/bash\necho 'hello'\n")
+				hookSidecarsValue := fmt.Sprintf(`[{"args": ["--version", "v1alpha2"], "image": "%s/%s:%s", "imagePullPolicy": "IfNotPresent"}]`,
+					flags.KubeVirtUtilityRepoPrefix, hookSidecarImage, flags.KubeVirtUtilityVersionTag)
+				vmi := libvmi.New(
+					libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+					libvmi.WithNetwork(v1.DefaultPodNetwork()),
+					libvmi.WithDataVolume("disk0", dv.Name),
+					libvmi.WithResourceMemory("1Gi"),
+					libvmi.WithCloudInitNoCloudEncodedUserData("#!/bin/bash\necho 'hello'\n"),
```
Comment (0xFelix) [inline tests/storage/hotplug.go:1323]: Same

File: tests/storage/imageupload.go [hunk 8]
```diff
@@ -28,13 +28,16 @@ import (
 	"kubevirt.io/client-go/kubecli"
 	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
```
Comment (0xFelix) [inline tests/storage/imageupload.go:30]: nit: newline

File: tests/storage/restore.go [hunk 9]
```diff
@@ -1834,10 +1834,17 @@ var _ = SIGDescribe("VirtualMachineRestore Tests", func() {
...
-					libstorage.AddDataVolumeTemplate(vm, dataVolume)
+					vmi := libvmi.New(
+						libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+						libvmi.WithNetwork(v1.DefaultPodNetwork()),
+						libvmi.WithDataVolume("disk0", dataVolume.Name),
+						libvmi.WithResourceMemory("1Gi"),
+						libvmi.WithCloudInitNoCloudEncodedUserData(bashHelloScript),
+						libvmi.WithNamespace(testsuite.GetTestNamespace(nil)),
+					)
+					vm := libvmi.NewVirtualMachine(vmi,
```
Comment (0xFelix) [inline tests/storage/restore.go:1845]: To inline or not to inline? :)

File: tests/migration/migration.go (hunk 3 above)
Comment (fossedihelm) [reply tests/migration/migration.go:2219]: I added a commit introducing a `BashHelloScript` constant under `cloudinit.go` file

File: tests/libstorage/datavolume.go (hunk 2 above)
Comment (fossedihelm) [reply tests/libstorage/datavolume.go:46]: Consider that it will be dropped in future PRs, since `AddDataVolume` and `AddDataVolumeTemplate` can be replaced by `libvmi` options.

File: tests/storage/datavolume.go (hunk 5 above)
Comment (fossedihelm) [reply tests/storage/datavolume.go]: Done

File: tests/storage/imageupload.go (hunk 8 above)
Comment (fossedihelm) [reply tests/storage/imageupload.go:30]: Done

**Comment by fossedihelm** [comment]
Thanks @0xFelix for the review! I addressed your comments. PTAL!

File: tests/storage/imageupload.go (hunk 8 above)
Comment (fossedihelm) [reply tests/storage/imageupload.go:24]: Ok this newline was added by `make && make generate`.
I re-arranged all the imports with some logic :)

File: tests/migration/migration.go (hunk 3 above)
Comment (0xFelix) [reply tests/migration/migration.go:2219]: But it's not used here?

File: tests/storage/datavolume.go (hunk 4 above)
Comment (0xFelix) [reply tests/storage/datavolume.go:1237]: Same, not used here?

File: tests/storage/storage.go [hunk 10]
```diff
@@ -80,7 +80,7 @@ const (
 
 	// Without cloud init user data Cirros takes long time to boot,
 	// so provide this dummy data to make it boot faster
-	cirrosUserData = "#!/bin/bash\necho 'hello'\n"
+	cirrosUserData = libvmi.BashHelloScript
```
Comment (0xFelix) [inline tests/storage/storage.go:83]: Can you inline it into uses of `cirrosUserData`?

File: tests/storage/datavolume.go (hunk 4 above)
Comment (fossedihelm) [reply tests/storage/datavolume.go]: Aaaaaaah that space between `\n` and `echo` :)
Fixed! Thank you

File: tests/storage/storage.go (hunk 10 above)
Comment (fossedihelm) [reply tests/storage/storage.go]: Dropped `cirrosUserData`

**Comment by fossedihelm** [comment]
> BashHelloScript

I checked :) And for what I found I've converted everything

File: tests/libvmi/cloudinit.go [hunk 11]
```diff
@@ -25,7 +25,10 @@ import (
 	v1 "kubevirt.io/api/core/v1"
 )
 
-const cloudInitDiskName = "disk1"
+const (
+	cloudInitDiskName = "disk1"
+	BashHelloScript   = "#!/bin/bash\necho 'hello'\n"
```
Comment (EdDev) [inline tests/libvmi/cloudinit.go:30]: I do not think this has anything to do with cloud-init configuration.
It is used for the Cirros and Alpine factories and should not be needed by anything else unless you show the need for it.

The usages for the Fedora based images are not needed and most likely can be dropped (they are leftovers).

**Review by EdDev — CHANGES_REQUESTED** [review]
I think the change made to `libvmi` are not needed.
The `echo hello` thing should not be exposed, it is an inner factory logic and should stay like that.

File: tests/libvmi/cloudinit.go (hunk 11 above)
Comment (fossedihelm) [reply tests/libvmi/cloudinit.go:30]: To me, it is widely used outside of Cirros and Alpine factories and it is always used with cloud-init configuration.
So from my point of view, for how things are right now, it fits the cloud-init file and it also deserves a unique const.
I also agree w/ u that in the long term, it can/will be dropped, but I think dropping a single place constant will be easier than dropping a plain spread string (with optional whitespace in the middle).
Having said that, if you all agree to leave things as is, I can easily drop the commit.
@EdDev @0xFelix please let me know!
Thanks

Comment (EdDev) [reply tests/libvmi/cloudinit.go:30]: Well, `libvmi/cloudinit.go` includes machinery for building cloudinit specs, not to keep there some odd cloud-init commands.

I prefer not to have it in `libvmi` for the reasons expressed before.

Comment (EdDev) [reply tests/libvmi/cloudinit.go:30]: I suspect the usages are leftovers of old code and that it is not needed anymore.

But if you prefer to investigate it in a follow-up, you are welcome to place stuff like this in some other lib*.

I need to move `libvmi` to `pkg/` so it can be used by unit tests without importing e2e packages.
If we start adding stuff like these in the package, we will interfere to that move.
But to be honest, even without this, I think it is wrong. That "hello" has no meaning and I do not think we should encourage contributors to use it.

Comment (fossedihelm) [reply tests/libvmi/cloudinit.go]: ok! Dropped the commit and rebased it onto main!
Thank you!

File: tests/migration/migration.go [hunk 12]
```diff
@@ -2197,8 +2212,13 @@ var _ = SIGMigrationDescribe("VM Live Migration", func() {
 				dv, err = virtClient.CdiClient().CdiV1beta1().DataVolumes(testsuite.GetTestNamespace(nil)).Create(context.Background(), dv, metav1.CreateOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				libstorage.EventuallyDV(dv, 600, HaveSucceeded())
-				vmi := tests.NewRandomVMIWithDataVolume(dv.Name)
-				tests.AddUserData(vmi, "disk1", "#!/bin/bash\n echo hello\n")
+				vmi := libvmi.New(
+					libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+					libvmi.WithNetwork(v1.DefaultPodNetwork()),
+					libvmi.WithDataVolume("disk0", dv.Name),
+					libvmi.WithResourceMemory("1Gi"),
```
Comment (EdDev) [inline tests/migration/migration.go:2219]: This is not really needed, together with the `hello` below.
The one who uses `newVirtualMachineInstanceWithFedoraRWXBlockDisk` will eventually override the memory anyway.

But it probably should be fixed in another commit or PR.

File: tests/migration/migration.go (hunk 3 above)
Comment (EdDev) [reply tests/migration/migration.go:2220]: It will also work without it, see comment above.

File: tests/storage/datavolume.go (hunk 4 above)
Comment (EdDev) [inline tests/storage/datavolume.go:1237]: Can be dropped as well, in a follow up.

**Review by EdDev — APPROVED** [review]
Thanks!

File: tests/libvmi/cloudinit.go (hunk 11 above)
Comment (EdDev) [reply tests/libvmi/cloudinit.go]: Yes, it is Cirros specific. This is why it is embedded in that factory.
 
 Most usages should use that factory.

Comment (0xFelix) [reply tests/libvmi/cloudinit.go]: What about cases where Cirros VMs using DVs are constructed manually? Like here? https://github.com/kubevirt/kubevirt/pull/11448

Comment (EdDev) [reply tests/libvmi/cloudinit.go]: I guess that is an exception and that factory is not relevant. For such exceptions I do not think `libvmi`needs to create a factory, that specific test can do it.

**Review by xpivarc — COMMENTED** [review]
/approve

**Comment by fossedihelm** [comment]
@0xFelix @EdDev Nothing changed, just rebased! PTAL
